### PR TITLE
Add llama.cpp GPU offload option

### PR DESCRIPTION
**Pull request by AlphaAtlas**
[requires #2058](https://github.com/oobabooga/text-generation-webui/pull/2058)

Splits models between GPU and CPU, see: https://github.com/ggerganov/llama.cpp/pull/1412

### CODE REVIEW

**Comment by AlphaAtlas** [comment]
Also, this should probably be tested on GPU-less llama builds (I can do this tonight), and a note about installing the GPU version of llama.cpp should be added: https://pypi.org/project/llama-cpp-python/

File: modules/llamacpp_model.py [hunk 1]
```diff
@@ -27,7 +27,8 @@ def from_pretrained(self, path):
             'n_threads': shared.args.threads or None,
             'n_batch': shared.args.n_batch,
             'use_mmap': not shared.args.no_mmap,
-            'use_mlock': shared.args.mlock
+            'use_mlock': shared.args.mlock,
+            'n_gpu_layers': shared.args.gpu_layers
```
Comment (mayaeary) [inline modules/llamacpp_model.py:31]:
```suggestion
            'n_gpu_layers': shared.args.n_gpu_layers
```

File: modules/shared.py [hunk 2]
```diff
@@ -123,6 +123,7 @@ def str2bool(v):
 parser.add_argument('--n_batch', type=int, default=512, help='Maximum number of prompt tokens to batch together when calling llama_eval.')
 parser.add_argument('--no-mmap', action='store_true', help='Prevent mmap from being used.')
 parser.add_argument('--mlock', action='store_true', help='Force the system to keep the model in RAM.')
+parser.add_argument('--n_gpu_layers', type=int, default=0, help='Number of layers to offload to the GPU.')
```
Comment (mayaeary) [inline modules/shared.py:126]:
```suggestion
parser.add_argument('--n-gpu-layers', type=int, default=0, help='Number of layers to offload to the GPU.')
```
To match llama.cpp argument - 
image

**Review by mayaeary — COMMENTED** [review]
A few suggestions.

**Comment by Malrama** [comment]
Now it's working! Perfect. Omg. This is brilliant!

**Comment by james-s-tayler** [comment]
This is miracle-tier.

**Comment by oobabooga** [comment]
It works for me. Not as fast as the old cuda branch gptq-for-llama yet but several times faster than cpu-only.

**Comment by skatardude10** [comment]
For anyone on the One-Click installer on Windows, I had to open a visual studio developer command prompt with build tools installed  (typed `cmake` to make sure I had it), navigate to my Oobabooga directory and run micromamba-cmd.bat to enter the venv,  and then typed

```
set CMAKE_ARGS="-DLLAMA_CUBLAS=on" 
set FORCE_CMAKE=1
```

to set those environment variables for the session, and then I was able to run successfully 

`pip install llama-cpp-python`

**Comment by LiliumSancta** [comment]
> It works for me. Not as fast as the old cuda branch gptq-for-llama yet but several times faster than cpu-only.
> 
> Documentation on the additional installation steps that are required:
> 
> https://github.com/oobabooga/text-generation-webui/blob/main/docs/llama.cpp-models.md#gpu-offloading

llama.cpp 

> It works for me. Not as fast as the old cuda branch gptq-for-llama yet but several times faster than cpu-only.
> 
> Documentation on the additional installation steps that are required:
> 
> https://github.com/oobabooga/text-generation-webui/blob/main/docs/llama.cpp-models.md#gpu-offloading

llama.cpp is my way to run 13B and larger models on my 8GB GPU, but there is a big difference in speed between running within text-generation-webui and running llama.cpp natively. I believe there is something wrong with the speed using the llama-cpp-python api, I don't know if it's something in the API itself or in the implementation, but in some cases the performance can be less than half of the original... See https: //github.com/abetlen/llama-cpp-python/issues/181

**Comment by marcostrada89** [comment]
> > > Come to think of it, maybe this should be fully automated? Like if an Nvidia GPU and build reqs are present, build llama cublast, otherwise build llama clblast if possible, otherwise default to the regular pip binary.
> > 
> > 
> > How do I do that on Windows? I have updated to 0.150 and everything is working fine but it doesn't seem to have cublast enabled. I tried using "CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python" but I am not 100% sure how I use this command in PowerShell. Do you know? :)
> 
> Try
> 
> ```
> pip cache remove llama-cpp-python
> pip uninstall llama-cpp-python
> CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python
> ```
> 
> llama-cpp-python won't rebuild if its in pip's cache.

I'm sorry, but i still didn't understand. I tried this command both in prompt and in python and it gives an error with "CMAKE_ARGS" as unrecognized.

Where or how is the code "CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python" supposed to be inserted? 


> For anyone on the One-Click installer on Windows, I had to open a visual studio developer command prompt with build tools installed (typed `cmake` to make sure I had it), navigate to my Oobabooga directory and run `micromamba-cmd.bat` to enter the venv, and then typed
> 
> ```
> set CMAKE_ARGS="-DLLAMA_CUBLAS=on" 
> set FORCE_CMAKE=1
> ```
> 
> to set those environment variables for the session, and then I was able to run successfully
> 
> `pip install llama-cpp-python`

I also tried this solution, but i can't find a micromamba-cmd.bat in all the folder.

I tried everything, both llama-cpp-python and LLAMA_CUBLAS were successfully installed, yet i can't offload anything to the gpu, i write the command line but the program seems to ignore it. Any help or hint would be appreciated, thank you.

**Comment by Thireus** [comment]
Does anyone have some performance benchmarks? I'm getting 5 tokens/s with q5_1 Vicuna 13B.

**Comment by AlphaAtlas** [comment]
@Thireus See: https://github.com/oobabooga/text-generation-webui/issues/2088

I havent actually gotten around to following up on this yet :P

**Comment by Smellbourne** [comment]
> > > > Come to think of it, maybe this should be fully automated? Like if an Nvidia GPU and build reqs are present, build llama cublast, otherwise build llama clblast if possible, otherwise default to the regular pip binary.
> > > 
> > > 
> > > How do I do that on Windows? I have updated to 0.150 and everything is working fine but it doesn't seem to have cublast enabled. I tried using "CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python" but I am not 100% sure how I use this command in PowerShell. Do you know? :)
> > 
> > 
> > Try
> > ```
> > pip cache remove llama-cpp-python
> > pip uninstall llama-cpp-python
> > CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > llama-cpp-python won't rebuild if its in pip's cache.
> 
> I'm sorry, but i still didn't understand. I tried this command both in prompt and in python and it gives an error with "CMAKE_ARGS" as unrecognized.
> 
> Where or how is the code "CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python" supposed to be inserted?
> 
> > For anyone on the One-Click installer on Windows, I had to open a visual studio developer command prompt with build tools installed (typed `cmake` to make sure I had it), navigate to my Oobabooga directory and run `micromamba-cmd.bat` to enter the venv, and then typed
> > ```
> > set CMAKE_ARGS="-DLLAMA_CUBLAS=on" 
> > set FORCE_CMAKE=1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to set those environment variables for the session, and then I was able to run successfully
> > `pip install llama-cpp-python`
> 
> I also tried this solution, but i can't find a micromamba-cmd.bat in all the folder.
> 
> I tried everything, both llama-cpp-python and LLAMA_CUBLAS were successfully installed, yet i can't offload anything to the gpu, i write the command line but the program seems to ignore it. Any help or hint would be appreciated, thank you.

Did you ever find a solution to this? I am in the exact same position.

**Comment by shikage** [comment]
> > > > > Come to think of it, maybe this should be fully automated? Like if an Nvidia GPU and build reqs are present, build llama cublast, otherwise build llama clblast if possible, otherwise default to the regular pip binary.
> > > > 
> > > > 
> > > > How do I do that on Windows? I have updated to 0.150 and everything is working fine but it doesn't seem to have cublast enabled. I tried using "CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python" but I am not 100% sure how I use this command in PowerShell. Do you know? :)
> > > 
> > > 
> > > Try
> > > ```
> > > pip cache remove llama-cpp-python
> > > pip uninstall llama-cpp-python
> > > CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > llama-cpp-python won't rebuild if its in pip's cache.
> > 
> > 
> > I'm sorry, but i still didn't understand. I tried this command both in prompt and in python and it gives an error with "CMAKE_ARGS" as unrecognized.
> > Where or how is the code "CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python" supposed to be inserted?
> > > For anyone on the One-Click installer on Windows, I had to open a visual studio developer command prompt with build tools installed (typed `cmake` to make sure I had it), navigate to my Oobabooga directory and run `micromamba-cmd.bat` to enter the venv, and then typed
> > > ```
> > > set CMAKE_ARGS="-DLLAMA_CUBLAS=on" 
> > > set FORCE_CMAKE=1
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > to set those environment variables for the session, and then I was able to run successfully
> > > `pip install llama-cpp-python`
> > 
> > 
> > I also tried this solution, but i can't find a micromamba-cmd.bat in all the folder.
> > I tried everything, both llama-cpp-python and LLAMA_CUBLAS were successfully installed, yet i can't offload anything to the gpu, i write the command line but the program seems to ignore it. Any help or hint would be appreciated, thank you.
> 
> Did you ever find a solution to this? I am in the exact same position.

You may be on windows, in which case try the approach from this earlier comment:
https://github.com/oobabooga/text-generation-webui/pull/2060#issuecomment-1547169518

**Pull request merged**